[DDP] Log some python-side errors

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59351 [DDP] log usage of torch_distributed_debug
* **#59284 [DDP] Log some python-side errors**
* #59281 [DDP] Log when errors happen

Logs a few python-side errors to DDP logging.

TODO: Most python errors actually have to do with user input correctness, so they throw before reducer is constructed and thus there is no logger. For this case, should we allow `logger` to be created optionally without a reducer, just for the purpose of logging errors, so that we can gain insight into these errors in scuba?

Differential Revision: [D28820290](https://our.internmc.facebook.com/intern/diff/D28820290/)

rohan-varma committed:
[DDP] Log some python-side errors

Logs a few python-side errors to DDP logging.

TODO: Most python errors actually have to do with user input correctness, so they throw before reducer is constructed and thus there is no logger. For this case, should we allow `logger` to be created optionally without a reducer, just for the purpose of logging errors, so that we can gain insight into these errors in scuba?

Differential Revision: [D28820290](https://our.internmc.facebook.com/intern/diff/D28820290/)

[ghstack-poisoned]

wayi1 reviewed (approved):
Thanks for the refactoring!

rohan-varma committed:
Update on "[DDP] Log some python-side errors"


Logs a few python-side errors to DDP logging.

TODO: Most python errors actually have to do with user input correctness, so they throw before reducer is constructed and thus there is no logger. For this case, should we allow `logger` to be created optionally without a reducer, just for the purpose of logging errors, so that we can gain insight into these errors in scuba?

Differential Revision: [D28820290](https://our.internmc.facebook.com/intern/diff/D28820290/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Log some python-side errors"


Logs a few python-side errors to DDP logging.

TODO: Most python errors actually have to do with user input correctness, so they throw before reducer is constructed and thus there is no logger. For this case, should we allow `logger` to be created optionally without a reducer, just for the purpose of logging errors, so that we can gain insight into these errors in scuba?

Differential Revision: [D28820290](https://our.internmc.facebook.com/intern/diff/D28820290/)

[ghstack-poisoned]
